exporter/elasticsearchexporter/exporter.go

```diff
@@ -579,19 +627,34 @@ func (e *elasticsearchExporter) pushProfileRecord(
 	profile pprofile.Profile,
 	defaultSession, eventsSession, stackTracesSession, stackFramesSession, executablesSession bulkIndexerSession,
 ) error {
-	return encoder.encodeProfile(ec, profile, func(buf *bytes.Buffer, docID, index string) error {
-		switch index {
-		case otelserializer.StackTraceIndex:
-			return stackTracesSession.Add(ctx, index, docID, "", buf, nil, docappender.ActionCreate)
-		case otelserializer.StackFrameIndex:
-			return stackFramesSession.Add(ctx, index, docID, "", buf, nil, docappender.ActionCreate)
-		case otelserializer.AllEventsIndex:
-			return eventsSession.Add(ctx, index, docID, "", buf, nil, docappender.ActionCreate)
-		case otelserializer.ExecutablesIndex:
-			return executablesSession.Add(ctx, index, docID, "", buf, nil, docappender.ActionUpdate)
-		default:
-			return defaultSession.Add(ctx, index, docID, "", buf, nil, docappender.ActionCreate)
-		}
+	return e.knownTraces.WithLock(func(tracesSet lru.LockedLRUSet[string]) error {
+		return e.knownFrames.WithLock(func(framesSet lru.LockedLRUSet[string]) error {
+			return e.knownExecutables.WithLock(func(executablesSet lru.LockedLRUSet[string]) error {
+				return encoder.encodeProfile(ec, profile, func(buf *bytes.Buffer, docID, index string) error {
+					switch index {
+					case otelserializer.StackTraceIndex:
+						if !tracesSet.CheckAndAdd(docID) {
```

Comment:
When we check the cache here, the JSON encoding already took place. So if docID is already in the cache, the encoding was for nothing and consumed resources unnecessarily.

Performance-wise, checking the caches in `transform.go` would avoid this kind overhead.

With that, you could also cache basic types like `FrameID`, `TraceID` and `FileID` instead of caching strings. Because strings are pointer types and counterfeit the "GC free" purpose of go-freelru: a cache instance with basic types is just a single block of memory, while with strings there are as many memory blocks as you inserted items.

Reply:
That would require a fairly large refactoring of the entire exporter, since only this struct here can hold state.
That's why I did it this way. But I can take the long road.

Reply:
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38732